client/state/signup/progress/reducer.js

```diff
@@ -0,0 +1,122 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { get, find, map, omit } from 'lodash';
+import debugFactory from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import stepsConfig from 'signup/config/steps';
+import flows from 'signup/config/flows';
+import {
+	SIGNUP_COMPLETE_RESET,
+	SIGNUP_PROGRESS_COMPLETE_STEP,
+	SIGNUP_PROGRESS_INVALIDATE_STEP,
+	SIGNUP_PROGRESS_PROCESS_STEP,
+	SIGNUP_PROGRESS_REMOVE_UNNEEDED_STEPS,
+	SIGNUP_PROGRESS_SAVE_STEP,
+	SIGNUP_PROGRESS_SET,
+	SIGNUP_PROGRESS_SUBMIT_STEP,
+} from 'state/action-types';
+import { createReducer } from 'state/utils';
+import { schema } from './schema';
+
+const debug = debugFactory( 'calypso:state:signup:progress:reducer' );
+
+//
+// Action handlers
+//
+function overwriteSteps( state, { steps = [] } ) {
+	return Array.isArray( steps ) ? steps : [];
+}
+
+// TODO: Write tests for this
+function completeStep( state, { step } ) {
+	return updateStep( { ...step, status: 'completed' } );
+}
+
+// TODO: Write tests for this
+function invalidateStep( state, { step, errors } ) {
+	const newStep = { ...step, errors };
+	if ( find( state, { stepName: newStep.stepName } ) ) {
+		return updateStep( state, newStep );
+	}
+	return addStep( state, { newStep } );
+}
+
+// TODO: Write tests for this
+export function processStep( state, { step } ) {
+	return updateStep( { ...step, status: 'processing' } );
+}
+
+function removeUnneededSteps( state, { flowName } ) {
+	const flowSteps = flows.getFlow( flowName ).steps;
+	return state.filter( step => flowSteps.includes( step.stepName ) );
+}
+
+// TODO: Write tests for this
+function saveStep( state, { step } ) {
+	if ( find( state, { stepName: step.stepName } ) ) {
+		return updateStep( state, step );
+	}
+
+	return addStep( state, { step, status: 'in-progress' } );
+}
+
+// TODO: Write tests for this
+function submitStep( state, { step } ) {
+	const stepHasApiRequestFunction = get( stepsConfig, `${ step.stepName }.apiRequestFunction` );
+	const status = stepHasApiRequestFunction ? 'pending' : 'completed';
+
+	return updateStep( { ...step, status } );
+}
+
+//
+// Helper Functions
+//
+function addStep( state, { step } ) {
+	debug( `Adding step ${ step.stepName }` );
+	return [ ...state, step ];
+}
+
+function updateStep( state, newStep ) {
```

Comment:
calls to this function appear to be passing one argument, the `step`, but this wants two

Reply:
Ah right, that's something I'm still fixing. They should all be fixed once the tests are up 👍